### ISSUE TITLE
rustdoc does not warn about broken links if they contain `.` or `[]`

### ISSUE THREAD

**Issue by whatisaphone** [rust-lang/rust#54191]
I got a PR yesterday that included something like this:

```rust
/// The tour de force of my career is [`foo.bar()`]
pub mod foo {
    pub fn bar() {}
}
```

In the current nightly rustdoc, `foo.bar()` neither resolves nor warns:

![](https://i.imgur.com/SZnon0r.png)

Our CI with `--deny warnings` did not catch this mistake (although I did 😄). I think rustdoc should trigger `intra-doc-link-resolution-failure` for this case.

**Comment by GuillaumeGomez** [comment]
The resolution failure is expected (should have been `foo::bar`) however, no warning is strange. I think it considers it as a path (as in a url) because of the `.`.

**Comment by QuietMisdreavus** [comment]
This is caused by this section of the intra-doc link code:

https://github.com/rust-lang/rust/blob/15d770400eed9018f18bddf83dd65cb7789280a5/src/librustdoc/passes/collect_intra_doc_links.rs#L356-L359

I think the idea here is that because of the dot, the link could be to a file, so it doesn't try to check it. (There's a similar skip earlier in the loop for if the link contains `/`.)

**Comment by jyn514** [comment]
@dylni mentioned in https://github.com/rust-lang/rust/issues/77199 that this also applies to things like `[&[String]]`, because of the brackets; it would be nice to warn about those. I know @Manishearth has been concerned in the past about warning too much in case it has false positives - maybe we could make this a new opt-in lint? as a bikeshed `warn(rustdoc::strange_intra_doc_links)` seems ok

**Comment by Manishearth** [comment]
I suspect a thing we can do is unconditionally hit the intra doc resolution path if it has backticks in it, at least. I'd rather not introduce another lint; we should work on improving the diagnostic instead.

**Comment by Nemo157** [comment]
Can we check whether it resolved as a markdown link too?

``[`foo.bar()`]`` is creating a link to an undefined reference `` `foo.bar()` ``, so should be warned about as a broken link whether it was intended to be an intra-doc link or not.

**Comment by jyn514** [comment]
@Nemo157 none of these links will ever resolve. `.` and `[]` are not paths and rustdoc doesn't special case them: https://github.com/rust-lang/rust/blob/8b954910c59a7a362c60959e93110892b6e9a691/src/librustdoc/passes/collect_intra_doc_links.rs#L2206-L2234

**Comment by jyn514** [comment]
Oh, you mean as an HTML link. Hmm, maybe? I have to look back over the pulldown API.

**Comment by Nemo157** [comment]
The `[foo]` form does not link to a URL, it references a URL defined elsewhere inside the markdown document using the identifier `foo`. If that identifier is not defined, I think technically it is not a link, it's just the text `[foo]`; but this is a syntax niche that intra-doc-links has already inserted itself into so I think it's fine for us to warn on all [_reference links_](https://spec.commonmark.org/0.30/#reference-link) that are missing a _link reference definition_ (including the implicit intra-doc-links definitions).

**Comment by Manishearth** [comment]
@Nemo157 you misunderstand; we only hit this codepath for _broken_ refs and for actual URLs. If it resolved to a markdown ref the ref title will never be seen here, though the URL passed to the ref title might (if it's sufficiently non-URLlike)

**Comment by Manishearth** [comment]
Oh, though I guess we could unconditionally warn on unresolving broken refs (but use the intra doc diagnostic on broken urls)

**Comment by lolbinarycat** [comment]
another case where these get confusing:

```
/// [whatever][1]
///
/// some other text
/// [1]: https://example.com/whatever
```

the `[1]:` does not actually get turned into a link reference, since there is no paragraph break before it.
